pg_search/src/postgres/storage/fsm.rs

```diff
@@ -34,413 +35,1263 @@ enum FSMBlockKind {
-    pub fn drain(
-        &mut self,
-        bman: &mut BufferManager,
-        n: usize,
-    ) -> impl Iterator<Item = pg_sys::BlockNumber> {
-        let xid_horizon =
-            unsafe { pg_sys::GetCurrentTransactionIdIfAny().max(pg_sys::FirstNormalTransactionId) };
-        let mut blocks = Vec::with_capacity(n);
-        let mut blockno = self.start_blockno;
-        loop {
-            if blockno == pg_sys::InvalidBlockNumber {
-                return blocks.into_iter();
+    /// When being drained, the FSM compares each block's stored `fsm_xid`` with this value, ensuring the stored
+    /// value precedes or equals this `xid_horizon`, before it is considered recyclable.
+    #[inline(always)]
+    fn passses_visibility_horizon(
+        fsm_xid: pg_sys::TransactionId,
+        xid_horizon: pg_sys::TransactionId,
+    ) -> bool {
+        crate::postgres::utils::TransactionIdPrecedesOrEquals(fsm_xid, xid_horizon)
+    }
+}
+
+/// The [`v2`] FreeSpaceManager is a fixed-size AVL tree laid out as an array on the FSM's first block.
+///
+/// Each entry in the AVL tree is called a [`Slot`] and each slot has a key, value, and tag.  The key
+/// is a 64bit [`pg_sys::FullTransactionId`], the value is actually the Rust unit type ([`()`]), and
+/// the tag is a [`pg_sys::BlockNumber`].
+///
+/// Each slot gets its own statically assigned tag value when a relation's FSM is first initialized.
+/// This value is the block number that represents the start of the list of block numbers that are
+/// free for the xid key stored in that slot.  This value stays with the slot, even as the tree is
+/// mutated as generally speaking, with an array-backed AVL tree entries don't move slots, only their
+/// left/right pointers are updated during rebalancing.  Special consideration is made for deleting
+/// an entry, however, when the key/value is moved to a new slot -- in this case the tag values are
+/// swapped until the entry is finished moving.  Essentially, tags have an affinity for their key/value
+/// as long as the slot is occupied.
+///
+/// The FSM has two user-facing operations: extend and drain.
+///
+/// # Extend
+///
+/// The caller provides both a transaction id and an iterator of free blocks to associate with that
+/// transaction.  That transaction indicates the point in time in which those blocks are usable.  Any
+/// future transaction `>=` to that transaction will be able to use those blocks.
+///
+/// During an extension a share lock is acquired on the FSM's root page (the AVL tree) and the
+/// specified transaction id is searched.  If it's found, then an exclusive lock is taken on the block
+/// represented by that xid's `tag` -- the blocklist -- and the blocklist is extended from head-to-tail,
+/// filling in gaps along the way.  However, if a full page is encountered, then a shortcut is taken
+/// and a new blocklist is linked into the existing list at that point.
```

Comment:
Nit: `blocklist` is a bit overloaded as a name in the codebase: maybe "freelist"?